### ISSUE TITLE
Make the topbar public by default 

### ISSUE THREAD

**Issue by AlexisTM** [meanjs/mean#583]
Dear,

I think we should make the topbar public by default or call the constructor elsewhere than in the service.

Else, the isPublic option is useless for EVERY menuItems or subMenuItems

~/modules/core/services/menus.client.service.js, #L172-175

        this.addMenu('topbar', {
            isPublic: true
        });

**Comment by ilanbiala** [comment]
@AlexisTM can you suggest what the code would look like and UX would be if it were public by default?

**Comment by AlexisTM** [comment]
Dear,

I did it on [Heroku](http://micd.herokuapp.com/) (free dyno to develop, so it can take 30 secs to come back from idle) and the [source is on Github](https://github.com/AlexisTM/MadeInWiki)

**THE PROBLEM :** 

To show the default menu (topbar) we have to modify the service. Indeed, if the `isPublic` property of the menu is set to `false`, every subMenus will be ignored even if the subMenu's `isPublic` property is set to `true`. [I suspect the HTML header `core/client/views/header.client.view.html` file for this.](https://github.com/AlexisTM/MadeInWiki/blob/df853e2370868d22323146ef4ff9ee3c4d2c9732/modules/core/client/views/header.client.view.html#L12-L25) If the menu's `shouldRender` returns false, the whole menu will be hidden by Angular.  This is *I think*, the desired behavior to hide a whole menu without worry to submenus. But the default menu is for *Noobs* like me who wants some menuItems to be visible to everyone and some menuItems to be visible to authenticated users only.

```HTML
<ul class="nav navbar-nav" data-ng-if="menu.shouldRender(authentication.user);">
    <li data-ng-repeat="item in menu.items | orderBy: 'position'" data-ng-if="item.shouldRender(authentication.user);" ng-switch="item.type" data-ng-class="{active: $state.includes(item.state)}" class="{{item.class}}" dropdown="item.type === 'dropdown'">
        <a ng-switch-when="dropdown" class="dropdown-toggle">
        <span data-ng-bind="item.title"></span>
        <b class="caret"></b>
        </a>
        <ul ng-switch-when="dropdown" class="dropdown-menu">
            <li data-ng-repeat="subitem in item.items | orderBy: 'position'" data-ng-if="subitem.shouldRender(authentication.user);" data-ui-sref-active="active">
                <a data-ui-sref="{{subitem.state}}" data-ng-bind="subitem.title"></a>
            </li>
        </ul>
        <a ng-switch-default data-ui-sref="{{item.state}}" data-ng-bind="item.title"></a>
    </li>
</ul>
```

**THE SOLUTION :**

The unique change is this one, when we create the menu. If we create the menu with isPublic : true, submenus will NEVER be rendered. [Code](https://github.com/AlexisTM/MadeInWiki/blob/master/modules/core/client/services/menus.client.service.js#L172-L175)

```javascript
        //Adding the topbar menu
        this.addMenu('topbar', {
            isPublic: true
        });
```

Changing the menu behavior force submenus to be rendered, every subMenus became `public`. for precaution, I added anyway the isPublic property to be clear. [Code](https://github.com/AlexisTM/MadeInWiki/blob/master/modules/articles/client/config/articles.client.config.js#L7-L18)

```javascript
		Menus.addMenuItem('topbar', {
			title: 'Articles',
			state: 'articles',
			type: 'dropdown',
			isPublic: true
		});

		// Add the dropdown list item
		Menus.addSubMenuItem('topbar', 'articles', {
			title: 'List Articles',
			state: 'articles.list'
		});
```

For every hidden menu, I added the `isPublic : false`. The result is only the article subMenu is shown. [Code](https://github.com/AlexisTM/MadeInWiki/blob/master/modules/categories/client/config/categories.client.config.js#L7-L18)

```javascript
		Menus.addMenuItem('topbar', {
			title: 'Categories',
			state: 'categories',
			type: 'dropdown',
			isPublic: false
		});

		// Add the dropdown list item
		Menus.addSubMenuItem('topbar', 'categories', {
			title: 'List Categories',
			state: 'categories.list'
		});
```

**Comment by danaronoff** [comment]
Thanks @AlexisTM, this helped me out.

**Comment by codydaig** [comment]
Is this something we want to change int he default boiler plate?

@rhutchison @lirantal @ilanbiala @mleanos @trainerbill @simison

**Comment by lirantal** [comment]
@codydaig I don't think it's an overall bad idea but I'm wondering if the issue raised by @AlexisTM  here is more than a decision and sounds more like a bug where it isn't easy to just set the menu visibility easily from the code.

**Comment by igorauad** [comment]
+1 @lirantal this is useful

**Comment by cdriscol** [comment]
:+1: I support this change in the boilerplate.  I've had to make this exact change in my projects.

**Comment by lirantal** [comment]
Ok so whose jumping the PR here? :)

**Comment by codydaig** [comment]
@lirantal I'll take this on.

**Comment by lirantal** [comment]
Great!

On Fri, Aug 7, 2015 at 3:22 AM, Cody B. Daig <notifications@github.com>
wrote:

> @lirantal <https://github.com/lirantal> I'll take this on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/meanjs/mean/issues/583#issuecomment-128547324>.
>



-- 
--
Sincerely, Liran Tal

Author
<http://www.amazon.com/Agile-Software-Development-HP-Manager/dp/1484210352/>of
Agile Software Development with HP Agile Manager
Founder and Lead Developer of daloRADIUS <http://www.daloradius.com/>
Blogging at http://www.enginx.com, and tweeting at @liran_tal
<https://twitter.com/liran_tal>

**Comment by codydaig** [comment]
Fixed in #780!

**Comment by AlexisTM** [comment]
I am glad I contributed to Mean!

**Comment by lirantal** [comment]
@AlexisTM keep it up  :)

**Comment by codydaig** [comment]
@lirantal I think this issue can be closed since it was resolved in #780.

**Comment by lirantal** [comment]
Thanks @codydaig
